File: services/python-ai/app/ai_legacy/middleware/principle_anchor.py
from typing import Any, Dict, Tuple

DEFAULT_PRINCIPLE = "Optimize for reconnection (clarity, continuity, coherence); avoid verbose comfort."


class PrincipleAnchorMiddleware:
    def __init__(self, enabled: bool = True, principle: str = DEFAULT_PRINCIPLE) -> None:
        self.config = {"enabled": enabled, "principle": principle}
        self._log: list[Dict[str, Any]] = []

    def inject_principle(self, system_prompt: str) -> str:
        if not self.config["enabled"]:
            return system_prompt
        return f"{system_prompt}\n\nNorth-Star Principle: {self.config['principle']}"

    def nudge_response(self, response: Any, context: Any) -> Dict[str, Any]:
        if not self.config["enabled"]:
            return {"nudged": response, "wasNudged": False}
        response_text = response if isinstance(response, str) else str(response)
        verbose_patterns = ["comfort", "it's okay", "don't worry", "take it easy", "relax"]
        has_verbose = any(pattern.lower() in response_text.lower() for pattern in verbose_patterns)
        reconnection_patterns = ["let's continue", "building on", "next step", "moving forward", "connect"]
        has_reconnection = any(pattern.lower() in response_text.lower() for pattern in reconnection_patterns)
        if has_verbose and not has_reconnection:
            nudge_reason = "Detected verbose comfort without reconnection focus"
            self._log.append({"timestamp": "", "action": "nudged", "reason": nudge_reason, "originalResponse": response})
            nudged_response = self._add_reconnection_element(response)
            return {"nudged": nudged_response, "wasNudged": True, "reason": nudge_reason}
        return {"nudged": response, "wasNudged": False}

    def _add_reconnection_element(self, response: Any) -> Any:
        if isinstance(response, str):
            return response + "\n\nLet's continue building on this foundation."
        return response

    def get_logs(self) -> list[Dict[str, Any]]:
        return list(self._log)

    def set_enabled(self, enabled: bool) -> None:
        self.config["enabled"] = enabled

    def get_config(self) -> Dict[str, Any]:
        return dict(self.config)


principle_anchor = PrincipleAnchorMiddleware()

File: apps/mobile/services/ai/middleware/principle_anchor.ts
export interface PrincipleAnchorConfig {
  enabled: boolean;
  principle: string;
}

const DEFAULT_PRINCIPLE = 'Optimize for reconnection (clarity, continuity, coherence); avoid verbose comfort.';

export class PrincipleAnchorMiddleware {
  private config: PrincipleAnchorConfig;
  private log: any[] = [];

  constructor(config: Partial<PrincipleAnchorConfig> = {}) {
    this.config = {
      enabled: config.enabled ?? true,
      principle: config.principle ?? DEFAULT_PRINCIPLE,
    };
  }

  // Inject principle into system prompt
  injectPrinciple(systemPrompt: string): string {
    if (!this.config.enabled) {
      return systemPrompt;
    }

    const principleInjection = `\n\nNorth-Star Principle: ${this.config.principle}`;
    return systemPrompt + principleInjection;
  }

  // Analyze response and nudge towards principle (without changing semantics)
  nudgeResponse(response: any, context: any): { nudged: any; wasNudged: boolean; reason?: string } {
    if (!this.config.enabled) {
      return { nudged: response, wasNudged: false };
    }

    // Analyze response content for principle alignment
    const responseText = typeof response === 'string' ? response : JSON.stringify(response);
    
    // Check for verbose comfort patterns
    const verbosePatterns = [
      /comfort/i,
      /it's okay/i,
      /don't worry/i,
      /take it easy/i,
      /relax/i,
    ];
    
    const hasVerboseComfort = verbosePatterns.some(pattern => pattern.test(responseText));
    
    // Check for reconnection patterns
    const reconnectionPatterns = [
      /let's continue/i,
      /building on/i,
      /next step/i,
      /moving forward/i,
      /connect/i,
    ];
    
    const hasReconnection = reconnectionPatterns.some(pattern => pattern.test(responseText));
    
    // Nudge logic: if verbose comfort detected and no reconnection, suggest reconnection
    if (hasVerboseComfort && !hasReconnection) {
      const nudgeReason = 'Detected verbose comfort without reconnection focus';
      this.log.push({
        timestamp: new Date().toISOString(),
        action: 'nudged',
        reason: nudgeReason,
        originalResponse: response,
      });
      
      // Add reconnection element without changing core content
      const nudgedResponse = this.addReconnectionElement(response);
      
      return {
        nudged: nudgedResponse,
        wasNudged: true,
        reason: nudgeReason,
      };
    }
    
    return { nudged: response, wasNudged: false };
  }

  private addReconnectionElement(response: any): any {
    // This is a placeholder - in practice, would modify response structure
    // to emphasize continuity/coherence without changing meaning
    if (typeof response === 'string') {
      return response + '\n\nLet\'s continue building on this foundation.';
    }
    return response; // For structured responses, would modify appropriately
  }

  // Get logs for A/B testing analysis
  getLogs(): any[] {
    return [...this.log];
  }

  // Toggle for A/B tests
  setEnabled(enabled: boolean): void {
    this.config.enabled = enabled;
  }

  getConfig(): PrincipleAnchorConfig {
    return { ...this.config };
  }
}

// Singleton instance for easy access
export const principleAnchor = new PrincipleAnchorMiddleware();
